### DISCUSSION TITLE
Additional HTTP response headers

### DISCUSSION THREAD

**Discussion by AstroViking** [directus/directus#6723] [Archive]
Hi all 👋 

Thank you for the great piece of software you are building! 

For our current project, we would be very happy to cache the requested item records (and related assets) on the client side (a mobile game programmed in Unity) without an explicit expiration date. However we'd also like them to get updated as soon as they change on the server.

As far as I know, this is not easily possible with Directus, since it does not set the `Last-Modified` or `ETag` response headers currently .

We are using Directus v9 at the moment, however it looks like there was a feature request for this for v8 (https://github.com/directus/v8-archive/issues/82). 

Is this something you still consider implementing?

On a further note, I also looked into implementing something like this in a custom extension. However the hooks found here (https://docs.directus.io/guides/api-hooks/) weren't really suited for the task.

- `items.read` => No possibility to set the headers
- `request` => Only exists as `request.not_found`, right?
- `response`  => Executed after the response has already been sent

Maybe one could hook into `middlewares.init.after` and register an additional middleware? Is this even a valid use case?

**Comment by rijkvanzanten** [comment]
> As far as I know, this is not easily possible with Directus, since it does not set the Last-Modified or ETag response headers currently .

We'll have to add a configuration option to directus_collections that marks a field as the "last updated" field (similar to `sort`) in order for the API to know what date to pass to the last-modified header 🤔

**Reply by benhaynes** [reply]
I like that approach, we're allowing users to set/map their own official usage fields. But curious, could we use `directus_activity` for this?

**Reply by rijkvanzanten** [reply]
@benhaynes we could, assuming the collection has accountability enabled 🤔

**Reply by RobinB-Gbanga** [reply]
Just chipping in to say that I would greatly appreciate this.

user-mappable fields would provide the greatest flexibility, but if it'd be easier as a first step, I think a lot of value could already come from allowing us to map either the default optional system field `date_updated`, if it is present on a table, or, as has been mentioned, the `directus_activity` (if that's not mixing responsibilities too much).

If well-documented, I could even imagine those two serving as default fallbacks, and projects/tables that offer neither of the two just don't provide a `Last-Modified` header.

In any case, I think this would add great value!

**Reply by philleepflorence** [reply]
### Follow Up Question
Will the D9 Event Hook `response.before` - *Feature Coming Soon* - function similarly to the D8 Filter Hook `response.GET` - to allow updating the response payload or injecting custom headers?

#### Previous Implementation
In D8, we leveraged the Collections: `directus_activity` + `directus_revisions` and the Filter Hook: `response.GET` to add Meta Properties to applicable JSON Response Payload.
#### D8 Example
```js
$payload.meta.<collection>.created = "action_on of first activity with action:create"
$payload.meta.<collection>.updated = "action_on of last activity with action:update"
```

**Reply by molszanski** [reply]
Implemented here:
https://github.com/directus/directus/pull/11818

**Comment by molszanski** [comment]
implemented here: https://github.com/directus/directus/pull/12231

**Comment by zhibirc** [comment]
What if the question is a bit more broader - do we have an ability in **Directus** ❤️ today, 2025-12-04, to set custom response headers? It's tightly related to security hardening of the Directus-based solutions. Thank you very much!

@rijkvanzanten FYI

**Reply by rijkvanzanten** [reply]
While I haven't tried this recently, you should be able to add a `middlewares.before` `init` hook to add in a new express middleware to set custom headers 🤔 

Something like (pseudo)

```ts
init('routes.before', (app) => {
  app.use((req, res, next) => {
    res.header('X-My-Custom-Header', 'SuperSecret')
  });
});
```
